{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from metadata_dict import (hca_keys, dcp_to_tier1_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dcp_path = \"./dcp_spreadsheets/\"\n",
    "flat_path = \"./flat_dcp/\"\n",
    "tier1_path = \"./tier1_output/\"\n",
    "file_name = \"ImmuneLandscapeccRCC_metadata_30-01-2023.xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2de639758adf4501b6fc124f41975b0d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Dropdown(description='Report Entity:', index=8, options=('RenalTumorMicroenvironment_metadata_06-04-2023.xlsx'…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ipywidgets as widgets\n",
    "from IPython.display import display\n",
    "file_name_dropdown = widgets.Dropdown(\n",
    "    options= os.listdir(path=dcp_path),\n",
    "    value=file_name or 'ImmuneLandscapeccRCC_metadata_30-01-2023.xlsx',\n",
    "    description='Report Entity:',\n",
    "    disabled=False,\n",
    ")\n",
    "display(file_name_dropdown)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Diabetic Nephropathy snRNA-seq_metadata_01-09-2023.xlsx\n"
     ]
    }
   ],
   "source": [
    "file_name = file_name_dropdown.value\n",
    "print(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "spreadsheet = pd.read_excel(os.path.join(dcp_path, file_name), sheet_name=None, skiprows= [0,1,2,4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Programmatic names as join keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save all biomaterials and protocols from each tab as join keys\n",
    "\n",
    "join_keys = {}\n",
    "for sheet in spreadsheet:\n",
    "    join_keys[sheet] = []\n",
    "    for key in spreadsheet[sheet].keys():\n",
    "        if (key.endswith(\"biomaterial_id\") or key.endswith(\"protocol_id\")) and \\\n",
    "                (not all(spreadsheet[sheet][key].isna())):\n",
    "            join_keys[sheet].append(key)\n",
    "    if not join_keys[sheet]:\n",
    "        del join_keys[sheet]\n",
    "\n",
    "\n",
    "join_set = set()\n",
    "for keys in join_keys.values():\n",
    "    join_set.update(keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save keys that have double pipe in order to merge metadata together for those keys\n",
    "\n",
    "keys_with_double_pipe = []\n",
    "for sheet in join_keys.keys():\n",
    "    for keys in join_keys[sheet]:\n",
    "        if any(spreadsheet[sheet][keys].dropna().str.contains(\"\\\\|\\\\|\")):\n",
    "            print(keys, sheet)\n",
    "            keys_with_double_pipe.append(keys)\n",
    "keys_with_double_pipe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spreadsheet tab order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Sequence file',\n",
       " 'Library preparation protocol',\n",
       " 'Sequencing protocol',\n",
       " 'Cell suspension',\n",
       " 'Dissociation protocol',\n",
       " 'Specimen from organism',\n",
       " 'Collection protocol',\n",
       " 'Donor organism']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# default ordering for simple experimental design & protocols\n",
    "# ordered_sheets = ['Analysis file', 'Cell suspension', 'Specimen from organism', 'Donor organism', \n",
    "#                  'Analysis protocol', 'Sequencing protocol', 'Library preparation protocol', 'Dissociation protocol', 'Enrichment protocol', 'Collection protocol']\n",
    "\n",
    "def field_id_to_tab(value):\n",
    "    \"\"\"Convert the programmatic name of a field to their entity/ tab name\"\"\"\n",
    "    return value.split('.')[0].replace('_', ' ').capitalize()\n",
    "\n",
    "\n",
    "ordered_sheets = []\n",
    "if 'Analysis file' in join_keys.keys():\n",
    "    tab = 'Analysis file'\n",
    "    ordered_sheets.append('Analysis file')\n",
    "    if 'Sequence file' in join_keys.keys():\n",
    "        ordered_sheets.append('Sequence file')\n",
    "else:\n",
    "    tab = 'Sequence file'\n",
    "    ordered_sheets.append('Sequence file')\n",
    "\n",
    "\n",
    "while len(ordered_sheets) < len(join_keys):\n",
    "    if tab == 'Donor organism':\n",
    "        break\n",
    "    for key in join_keys[tab]:\n",
    "        if field_id_to_tab(key) in ordered_sheets:\n",
    "            continue\n",
    "        if key.endswith('protocol_id'):\n",
    "            ordered_sheets.append(field_id_to_tab(key))\n",
    "        elif key.endswith('biomaterial_id'):\n",
    "            tab = field_id_to_tab(key)\n",
    "            ordered_sheets.append(tab)\n",
    "\n",
    "ordered_sheets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'Analysis file' in spreadsheet:\n",
    "    # filter out analysis files that will create duplicates i.e. barcodes and features for same CS and remove analysis_files that did not derive from CS\n",
    "    spreadsheet['Analysis file'] = spreadsheet['Analysis file'][\\\n",
    "        spreadsheet['Analysis file']['analysis_file.file_core.content_description.ontology'].str.contains('data:3917|data:3112',na=False)]\\\n",
    "        .dropna(subset='cell_suspension.biomaterial_core.biomaterial_id')\n",
    "    \n",
    "    # If we have pooled CS for analysis_files, create a new entry for each CS of each file\n",
    "    analysis_pooledCS = spreadsheet['Analysis file']['cell_suspension.biomaterial_core.biomaterial_id'].str.contains(\"\\\\|\\\\|\")\n",
    "    if any(analysis_pooledCS):\n",
    "        print(f\"analysis files are pooled\")\n",
    "        spreadsheet['Analysis file']['cell_suspension.biomaterial_core.biomaterial_id'] = \\\n",
    "            spreadsheet['Analysis file']['cell_suspension.biomaterial_core.biomaterial_id']\\\n",
    "                .str.split('\\\\|\\\\|')\\\n",
    "                .apply(lambda x: x[0] if len(x) == 1 else x)\n",
    "        spreadsheet['Analysis file'] = spreadsheet['Analysis file'].explode('cell_suspension.biomaterial_core.biomaterial_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit specific process field names to their Tier 1 field since there are multiple process fields in each sheet\n",
    "\n",
    "process_fields = {\n",
    "    'Specimen from organism': {'process.process_core.location': 'sample_collection_site'},\n",
    "    'Sequence file': {'process.process_core.process_id': 'library_sequencing_run',\n",
    "                      'process.insdc_experiment.insdc_experiment_accession': 'library_id_repository'}\n",
    "}\n",
    "\n",
    "for sheet in process_fields:\n",
    "    if sheet not in spreadsheet:\n",
    "        continue\n",
    "    spreadsheet[sheet] = spreadsheet[sheet].rename(columns=process_fields[sheet])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Select the library preparations that you want to include ['library_prep_protocol']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def collapse_values(series):\n",
    "    return \", \".join(series.unique().astype(str))\n",
    "    \n",
    "# If we have insdc run accessions, and custom process_id includes this accession, then use insdc run accession instead of process_id\n",
    "if 'Sequence file' in spreadsheet and \\\n",
    "    'sequence_file.insdc_run_accessions' in spreadsheet['Sequence file'] and \\\n",
    "        spreadsheet['Sequence file']['sequence_file.insdc_run_accessions'].notna().any():\n",
    "    insdc_in_process = spreadsheet['Sequence file'].apply(lambda row: row['sequence_file.insdc_run_accessions'] in row['library_sequencing_run'], axis=1)\n",
    "    spreadsheet['Sequence file'].loc[insdc_in_process, 'library_sequencing_run'] = spreadsheet['Sequence file'].loc[insdc_in_process, 'sequence_file.insdc_run_accessions']\n",
    "\n",
    "# Remove files from library_preparation_protocol that has to do with modality that will not be included in the count_matrix\n",
    "if 'Sequence file' in spreadsheet:\n",
    "    print(f\"Select the library preparations that you want to include {spreadsheet['Sequence file']['library_preparation_protocol.protocol_core.protocol_id'].unique()}\\n\")\n",
    "    if True:\n",
    "        # Exclude this library_preparation_protocol <SELECT HERE>\n",
    "        exclude_lib_prep = ['TCR_library_prep']\n",
    "        spreadsheet['Sequence file'] = spreadsheet['Sequence file'][~spreadsheet['Sequence file']['library_preparation_protocol.protocol_core.protocol_id'].isin(exclude_lib_prep)]\n",
    "    if False:\n",
    "        # Include this library_preparation_protocol <OR SELECT HERE>\n",
    "        include_lib_prep = ['10x_library_prep']\n",
    "        spreadsheet['Sequence file'] = spreadsheet['Sequence file'][spreadsheet['Sequence file']['library_preparation_protocol.protocol_core.protocol_id'].isin(include_lib_prep)]\n",
    "\n",
    "    # From sequence file we record everything at the CS level\n",
    "    spreadsheet['Sequence file'] = spreadsheet['Sequence file']\\\n",
    "        .groupby('cell_suspension.biomaterial_core.biomaterial_id')\\\n",
    "        .agg(collapse_values)\\\n",
    "        .reset_index()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialise output dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initiate flat dataframe with analysis_files and unique cell_suspensions (bc for barcode)\n",
    "if 'Analysis file' in spreadsheet:\n",
    "    bc = spreadsheet['Analysis file'][['cell_suspension.biomaterial_core.biomaterial_id','analysis_file.file_core.file_name']]\\\n",
    "        .drop_duplicates('cell_suspension.biomaterial_core.biomaterial_id')\n",
    "\n",
    "# If no analysis files available provide flatten metatadata at the sample level based on the CS from the sequence file tab\n",
    "# TODO unpool CS for sequence files too\n",
    "elif 'Sequence file' in spreadsheet:\n",
    "    bc = spreadsheet['Sequence file'].loc[:, ['cell_suspension.biomaterial_core.biomaterial_id']]\\\n",
    "        .drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if we have pooled keys, edit the spreadsheet and add the pooled values as extra rows\n",
    "\n",
    "def process_pooled_keys(sheet, bc, join_key):\n",
    "    pooled_keys = [key for key in bc[join_key].dropna().unique().tolist() if \"||\" in key]\n",
    "    print(\"We have pooled join keys\", pooled_keys)\n",
    "    for key in pooled_keys:\n",
    "        unpooled_bool = sheet[join_key].isin(key.split(\"||\"))\n",
    "        summary_row = {}\n",
    "        unique_values = {}\n",
    "        for col, contents in sheet.loc[unpooled_bool].items():\n",
    "            unique_values[col] = list(contents.dropna().astype('str').unique())\n",
    "            if len(unique_values[col]) == 1:\n",
    "                unique_values[col] = unique_values[col][0]\n",
    "        for col, values in unique_values.items():\n",
    "            if isinstance(values, list) and len(values) > 1:\n",
    "                summary_row[col] = '||'.join(values)\n",
    "            else:\n",
    "                summary_row[col] = values\n",
    "        id_key = [name for name in summary_row.keys() if name.endswith(\"_id\")]\n",
    "        if len(id_key) > 1:\n",
    "            print(f\"Multiple id keys {', '.join(id_key)}\")\n",
    "        summary_row[id_key[0]] = key\n",
    "        return pd.concat([sheet, pd.DataFrame([summary_row])], ignore_index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging sheet Sequence file on key cell_suspension.biomaterial_core.biomaterial_id\n",
      "Shape is (6, 18)\n",
      "Merging sheet Library preparation protocol on key library_preparation_protocol.protocol_core.protocol_id\n",
      "Shape is (6, 37)\n",
      "Merging sheet Sequencing protocol on key sequencing_protocol.protocol_core.protocol_id\n",
      "Shape is (6, 47)\n",
      "Merging sheet Cell suspension on key cell_suspension.biomaterial_core.biomaterial_id\n",
      "Shape is (6, 64)\n",
      "Merging sheet Dissociation protocol on key dissociation_protocol.protocol_core.protocol_id\n",
      "Shape is (6, 70)\n",
      "Merging sheet Specimen from organism on key specimen_from_organism.biomaterial_core.biomaterial_id\n",
      "Shape is (6, 93)\n",
      "Merging sheet Collection protocol on key collection_protocol.protocol_core.protocol_id\n",
      "Shape is (6, 98)\n",
      "Merging sheet Donor organism on key donor_organism.biomaterial_core.biomaterial_id\n",
      "Shape is (6, 119)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_suspension.biomaterial_core.biomaterial_id</th>\n",
       "      <th>sequence_file.uuid</th>\n",
       "      <th>sequence_file.file_core.file_name</th>\n",
       "      <th>sequence_file.file_core.format</th>\n",
       "      <th>sequence_file.file_core.content_description.text</th>\n",
       "      <th>sequence_file.file_core.content_description.ontology</th>\n",
       "      <th>sequence_file.file_core.content_description.ontology_label</th>\n",
       "      <th>sequence_file.file_core.checksum</th>\n",
       "      <th>sequence_file.read_index</th>\n",
       "      <th>sequence_file.library_prep_id</th>\n",
       "      <th>...</th>\n",
       "      <th>donor_organism.organism_age</th>\n",
       "      <th>donor_organism.development_stage.text</th>\n",
       "      <th>donor_organism.development_stage.ontology</th>\n",
       "      <th>donor_organism.development_stage.ontology_label</th>\n",
       "      <th>donor_organism.diseases.text</th>\n",
       "      <th>donor_organism.diseases.ontology</th>\n",
       "      <th>donor_organism.diseases.ontology_label</th>\n",
       "      <th>donor_organism.medical_history.nutritional_state</th>\n",
       "      <th>donor_organism.medical_history.test_results</th>\n",
       "      <th>donor_organism.medical_history.medication</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cell_suspension_control_1</td>\n",
       "      <td>51c04a77-cb8e-492a-83e9-de9c46c37729, 2a9e7377...</td>\n",
       "      <td>R1.fastq.gz, R2.fastq.gz</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>f9491a935fb91f5463cfd65b6f9203a0, 36d0912fc6e8...</td>\n",
       "      <td>read1, read2</td>\n",
       "      <td>library1</td>\n",
       "      <td>...</td>\n",
       "      <td>54</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>Renal mass||Acoustic neuroma||OSA</td>\n",
       "      <td>MONDO:0002367||MONDO:0001569||MONDO:0007147</td>\n",
       "      <td>kidney cancer||acoustic neuroma||obstructive s...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>sCr 1.28 mg/dL, serum K+ 3.9  mEq/L</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell_suspension_control_2</td>\n",
       "      <td>336597e4-b3af-4ab9-8425-ea9f1c04799b, f7204ad4...</td>\n",
       "      <td>Norm_S3_L001_R1_001.fastq.gz, Norm_S3_L001_R2_...</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>31d6d97bf19cc00cfc6313178af554b0, 1fd575225402...</td>\n",
       "      <td>read1, read2, index1</td>\n",
       "      <td>library2</td>\n",
       "      <td>...</td>\n",
       "      <td>62</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>Renal mass||Hypertension||AKI (5 months prior)...</td>\n",
       "      <td>MONDO:0002367||MONDO:0001134||MONDO:0006637||M...</td>\n",
       "      <td>kidney cancer||essential hypertension||acute k...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>sCr 1.21 mg/dL, serum K+ 3.3  mEq/L</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cell_suspension_control_3</td>\n",
       "      <td>9c2c35a2-1400-4f1e-b870-750dbebf6b63, f0e4f00f...</td>\n",
       "      <td>control_S2_L001_R1_001.fastq.gz, control_S2_L0...</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>8c1ecc2584e5e2f1e07ae97ef0b6d44d, a51d6351b220...</td>\n",
       "      <td>read1, read2, index1</td>\n",
       "      <td>library3</td>\n",
       "      <td>...</td>\n",
       "      <td>61</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>Renal mass||anxiety||depression||arthritis||di...</td>\n",
       "      <td>MONDO:0002367||MONDO:0005618||MONDO:0002050||M...</td>\n",
       "      <td>kidney cancer||anxiety disorder||depressive di...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>sCr 0.89 mg/dL, serum K+ 3.8  mEq/L</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell_suspension_diabetes_1</td>\n",
       "      <td>e0ca126d-7d50-48b9-bfc3-18a419d92499, 880de2e7...</td>\n",
       "      <td>DN.R1.fastq.gz, DN.R2.fastq.gz</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>8233b866f89a36e2f6a3b10bf54aba46, 399c187b88aa...</td>\n",
       "      <td>read1, read2</td>\n",
       "      <td>library4</td>\n",
       "      <td>...</td>\n",
       "      <td>74</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>right renal mass||type 2 diabetes||adrenal mas...</td>\n",
       "      <td>MONDO:0002367||MONDO:0005148||MONDO:0003924||M...</td>\n",
       "      <td>kidney cancer||type 2 diabetes mellitus||adren...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>A1c 7.3 mmol/mol, sCr 1.08 mg/dL, serum K+ 4.2...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cell_suspension_diabetes_2</td>\n",
       "      <td>11e0bb39-9abc-4e8a-906f-73bc5d89e66b, eafbf1ec...</td>\n",
       "      <td>DN_S2_L001_R1_001.fastq.gz, DN_S2_L001_R2_001....</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>52f6ef9f80ecf909e60672d060444ab8, 34f0c79389fe...</td>\n",
       "      <td>read1, read2, index1</td>\n",
       "      <td>library5</td>\n",
       "      <td>...</td>\n",
       "      <td>52</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>Renal mass||Hypertension||type 2 diabetes||pro...</td>\n",
       "      <td>MONDO:0002367||MONDO:0001134||MONDO:0005148||M...</td>\n",
       "      <td>kidney cancer||essential hypertension||type 2 ...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>A1c 6.7 mmol/mol, sCr 1.26 mg/dL, serum K+ 4.3...</td>\n",
       "      <td>ACE inhibitor, HCTZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cell_suspension_diabetes_3</td>\n",
       "      <td>7f495d31-78db-46a8-aa3b-e399913b0118, 8518a4e5...</td>\n",
       "      <td>DN2_S1_L001_R1_001.fastq.gz, DN2_S1_L001_R2_00...</td>\n",
       "      <td>fastq.gz</td>\n",
       "      <td>DNA sequence (raw)</td>\n",
       "      <td>data:3494</td>\n",
       "      <td>DNA sequence</td>\n",
       "      <td>a34bbda366fe4b35c380ef768b8c1def, 7d94c4c2ace8...</td>\n",
       "      <td>read1, read2, index1</td>\n",
       "      <td>library6</td>\n",
       "      <td>...</td>\n",
       "      <td>57</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>HsapDv:0000087</td>\n",
       "      <td>human adult stage</td>\n",
       "      <td>Renal mass||type 2 diabetes||GERD||hyperlipidemia</td>\n",
       "      <td>MONDO:0002367||MONDO:0005148||MONDO:0007186||M...</td>\n",
       "      <td>kidney cancer||type 2 diabetes mellitus||gastr...</td>\n",
       "      <td>fasting</td>\n",
       "      <td>A1c 9.7 mmol/mol, sCr 0.7 mg/dL, serum K+ 4.2 ...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 119 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_suspension.biomaterial_core.biomaterial_id  \\\n",
       "0                       cell_suspension_control_1   \n",
       "1                       cell_suspension_control_2   \n",
       "2                       cell_suspension_control_3   \n",
       "3                      cell_suspension_diabetes_1   \n",
       "4                      cell_suspension_diabetes_2   \n",
       "5                      cell_suspension_diabetes_3   \n",
       "\n",
       "                                  sequence_file.uuid  \\\n",
       "0  51c04a77-cb8e-492a-83e9-de9c46c37729, 2a9e7377...   \n",
       "1  336597e4-b3af-4ab9-8425-ea9f1c04799b, f7204ad4...   \n",
       "2  9c2c35a2-1400-4f1e-b870-750dbebf6b63, f0e4f00f...   \n",
       "3  e0ca126d-7d50-48b9-bfc3-18a419d92499, 880de2e7...   \n",
       "4  11e0bb39-9abc-4e8a-906f-73bc5d89e66b, eafbf1ec...   \n",
       "5  7f495d31-78db-46a8-aa3b-e399913b0118, 8518a4e5...   \n",
       "\n",
       "                   sequence_file.file_core.file_name  \\\n",
       "0                           R1.fastq.gz, R2.fastq.gz   \n",
       "1  Norm_S3_L001_R1_001.fastq.gz, Norm_S3_L001_R2_...   \n",
       "2  control_S2_L001_R1_001.fastq.gz, control_S2_L0...   \n",
       "3                     DN.R1.fastq.gz, DN.R2.fastq.gz   \n",
       "4  DN_S2_L001_R1_001.fastq.gz, DN_S2_L001_R2_001....   \n",
       "5  DN2_S1_L001_R1_001.fastq.gz, DN2_S1_L001_R2_00...   \n",
       "\n",
       "  sequence_file.file_core.format  \\\n",
       "0                       fastq.gz   \n",
       "1                       fastq.gz   \n",
       "2                       fastq.gz   \n",
       "3                       fastq.gz   \n",
       "4                       fastq.gz   \n",
       "5                       fastq.gz   \n",
       "\n",
       "  sequence_file.file_core.content_description.text  \\\n",
       "0                               DNA sequence (raw)   \n",
       "1                               DNA sequence (raw)   \n",
       "2                               DNA sequence (raw)   \n",
       "3                               DNA sequence (raw)   \n",
       "4                               DNA sequence (raw)   \n",
       "5                               DNA sequence (raw)   \n",
       "\n",
       "  sequence_file.file_core.content_description.ontology  \\\n",
       "0                                          data:3494     \n",
       "1                                          data:3494     \n",
       "2                                          data:3494     \n",
       "3                                          data:3494     \n",
       "4                                          data:3494     \n",
       "5                                          data:3494     \n",
       "\n",
       "  sequence_file.file_core.content_description.ontology_label  \\\n",
       "0                                       DNA sequence           \n",
       "1                                       DNA sequence           \n",
       "2                                       DNA sequence           \n",
       "3                                       DNA sequence           \n",
       "4                                       DNA sequence           \n",
       "5                                       DNA sequence           \n",
       "\n",
       "                    sequence_file.file_core.checksum sequence_file.read_index  \\\n",
       "0  f9491a935fb91f5463cfd65b6f9203a0, 36d0912fc6e8...             read1, read2   \n",
       "1  31d6d97bf19cc00cfc6313178af554b0, 1fd575225402...     read1, read2, index1   \n",
       "2  8c1ecc2584e5e2f1e07ae97ef0b6d44d, a51d6351b220...     read1, read2, index1   \n",
       "3  8233b866f89a36e2f6a3b10bf54aba46, 399c187b88aa...             read1, read2   \n",
       "4  52f6ef9f80ecf909e60672d060444ab8, 34f0c79389fe...     read1, read2, index1   \n",
       "5  a34bbda366fe4b35c380ef768b8c1def, 7d94c4c2ace8...     read1, read2, index1   \n",
       "\n",
       "  sequence_file.library_prep_id  ... donor_organism.organism_age  \\\n",
       "0                      library1  ...                          54   \n",
       "1                      library2  ...                          62   \n",
       "2                      library3  ...                          61   \n",
       "3                      library4  ...                          74   \n",
       "4                      library5  ...                          52   \n",
       "5                      library6  ...                          57   \n",
       "\n",
       "  donor_organism.development_stage.text  \\\n",
       "0                     human adult stage   \n",
       "1                     human adult stage   \n",
       "2                     human adult stage   \n",
       "3                     human adult stage   \n",
       "4                     human adult stage   \n",
       "5                     human adult stage   \n",
       "\n",
       "  donor_organism.development_stage.ontology  \\\n",
       "0                            HsapDv:0000087   \n",
       "1                            HsapDv:0000087   \n",
       "2                            HsapDv:0000087   \n",
       "3                            HsapDv:0000087   \n",
       "4                            HsapDv:0000087   \n",
       "5                            HsapDv:0000087   \n",
       "\n",
       "  donor_organism.development_stage.ontology_label  \\\n",
       "0                               human adult stage   \n",
       "1                               human adult stage   \n",
       "2                               human adult stage   \n",
       "3                               human adult stage   \n",
       "4                               human adult stage   \n",
       "5                               human adult stage   \n",
       "\n",
       "                        donor_organism.diseases.text  \\\n",
       "0                  Renal mass||Acoustic neuroma||OSA   \n",
       "1  Renal mass||Hypertension||AKI (5 months prior)...   \n",
       "2  Renal mass||anxiety||depression||arthritis||di...   \n",
       "3  right renal mass||type 2 diabetes||adrenal mas...   \n",
       "4  Renal mass||Hypertension||type 2 diabetes||pro...   \n",
       "5  Renal mass||type 2 diabetes||GERD||hyperlipidemia   \n",
       "\n",
       "                    donor_organism.diseases.ontology  \\\n",
       "0        MONDO:0002367||MONDO:0001569||MONDO:0007147   \n",
       "1  MONDO:0002367||MONDO:0001134||MONDO:0006637||M...   \n",
       "2  MONDO:0002367||MONDO:0005618||MONDO:0002050||M...   \n",
       "3  MONDO:0002367||MONDO:0005148||MONDO:0003924||M...   \n",
       "4  MONDO:0002367||MONDO:0001134||MONDO:0005148||M...   \n",
       "5  MONDO:0002367||MONDO:0005148||MONDO:0007186||M...   \n",
       "\n",
       "              donor_organism.diseases.ontology_label  \\\n",
       "0  kidney cancer||acoustic neuroma||obstructive s...   \n",
       "1  kidney cancer||essential hypertension||acute k...   \n",
       "2  kidney cancer||anxiety disorder||depressive di...   \n",
       "3  kidney cancer||type 2 diabetes mellitus||adren...   \n",
       "4  kidney cancer||essential hypertension||type 2 ...   \n",
       "5  kidney cancer||type 2 diabetes mellitus||gastr...   \n",
       "\n",
       "  donor_organism.medical_history.nutritional_state  \\\n",
       "0                                          fasting   \n",
       "1                                          fasting   \n",
       "2                                          fasting   \n",
       "3                                          fasting   \n",
       "4                                          fasting   \n",
       "5                                          fasting   \n",
       "\n",
       "         donor_organism.medical_history.test_results  \\\n",
       "0                sCr 1.28 mg/dL, serum K+ 3.9  mEq/L   \n",
       "1                sCr 1.21 mg/dL, serum K+ 3.3  mEq/L   \n",
       "2                sCr 0.89 mg/dL, serum K+ 3.8  mEq/L   \n",
       "3  A1c 7.3 mmol/mol, sCr 1.08 mg/dL, serum K+ 4.2...   \n",
       "4  A1c 6.7 mmol/mol, sCr 1.26 mg/dL, serum K+ 4.3...   \n",
       "5  A1c 9.7 mmol/mol, sCr 0.7 mg/dL, serum K+ 4.2 ...   \n",
       "\n",
       "  donor_organism.medical_history.medication  \n",
       "0                                       NaN  \n",
       "1                                       NaN  \n",
       "2                                       NaN  \n",
       "3                                       NaN  \n",
       "4                       ACE inhibitor, HCTZ  \n",
       "5                                       NaN  \n",
       "\n",
       "[6 rows x 119 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# in each sheet merge the flat df with the spreadsheet\n",
    "\n",
    "for sheet in ordered_sheets:\n",
    "    if sheet not in spreadsheet.keys():\n",
    "        continue\n",
    "    keys_in_tab = spreadsheet[sheet].keys()\n",
    "    join_key = [key for key in keys_in_tab if key in bc.keys() and key in join_set]\n",
    "    if len(join_key) > 1:\n",
    "        print(\"Multiple join keys in \" + sheet + \":\\n\\t\" + \"\\n\\t\".join(join_key))\n",
    "        if sheet in ['Analysis file', 'Sequence file']:\n",
    "            join_key = 'cell_suspension.biomaterial_core.biomaterial_id'\n",
    "    else:\n",
    "        join_key = join_key[0]\n",
    "\n",
    "    keys_in_tab = [key for key in keys_in_tab if key not in bc.keys() or key == join_key]\n",
    "    if join_key in keys_with_double_pipe and any(bc[join_key].str.contains(\"\\\\|\\\\|\")):\n",
    "        spreadsheet[sheet] = process_pooled_keys(spreadsheet[sheet], bc, join_key)\n",
    "\n",
    "    print(\"Merging sheet \" + sheet + \" on key \" + join_key, sep = \"\\t\")\n",
    "    bc = bc.merge(spreadsheet[sheet][keys_in_tab], on = join_key, how = \"left\")\n",
    "    print(\"Shape is \" + str(bc.shape))\n",
    "bc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "bc.to_csv(os.path.join(flat_path, file_name.replace(\".xlsx\", \"_flat_biomaterial.csv\")), index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ingest_api",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
